### ISSUE TITLE
Levelled enrichers/`LoggerEnrichmentConfiguration.Wrap()`

### ISSUE THREAD

**Issue by nblumhardt** [serilog/serilog#1310]
> @nblumhardt Is there a way to enable an enricher only when a specific severity level is enabled? It would be nice to be able to use this code only when, say, Verbose is enabled.

_Originally posted by @andre-ss6 in https://github.com/serilog/serilog/issues/1084#issuecomment-492396008_

**Comment by nblumhardt** [comment]
It would be nice to go back in time and build levelling support into enricher configuration methods, much like we've done with sinks, but there's a lot of churn involved in that.

Alternatively, we could make it possible for a levelled enricher wrapper to be written:

```csharp
    .Enrich.AtLevel(LogEventLevel.Verbose, e => e.WithMachineName())
```

We provide `LoggerSinkConfiguration.Wrap()` so that wrapper sinks can modify the behavior of other sinks, and `Enrich.AtLevel()` could be built on a similar `LoggerEnrichmentConfiguration.Wrap()` method, were one available.

Thoughts? Are there any other use cases that could be implemented on top of this? Would a similar method for destructuring also find use?

**Comment by vlm---** [comment]
+1 for leveled enrichment.
Leveled approach would be also helpful for destructuring, omitting certain properties on higher levels but adding them on others. It should support including or excluding configuration with list of levels and higher / lower than level approach.
In one of our projects we used forked [Destructurama.Attributed](https://github.com/destructurama/attributed) with added attribute `LogAtAttribute(LogEventLevel level)` for ignoring properties below defined level, but performance of that implementation was not optimal for widespread use and we ditched it after the project matured and verbose /debug logging was not needed as much.

**Comment by nblumhardt** [comment]
Just looking at the example again, we'd need to use `LoggingLevelSwitch` with the API if we want to avoid having to re-specify the minimum level:

```csharp
    .Enrich.AtLevel(LogEventLevel.Verbose, levelSwitch, e => e.WithMachineName())
```

**Comment by abrasat** [comment]
> In one of our projects we used forked [Destructurama.Attributed](https://github.com/destructurama/attributed) with added attribute `LogAtAttribute(LogEventLevel level)` for ignoring properties below defined level, but performance of that implementation was not optimal for widespread use and we ditched it after the project matured and verbose /debug logging was not needed as much.

Could you give some details about bad performance issues ?

Is it basically a "bad" idea to use serilog enrichers, as these impact on application performance?

**Comment by vlm---** [comment]
@nblumhardt this could lead to some confusion when levelSwitch instance is not used to control actual logger we are configuring. And one more point:
`.Enrich.AtLevel(LogEventLevel.Debug, levelSwitch, e => e.WithMachineName())`
means that it works for Verbose as well or only for Debug?

@abrasat Every code that has to run does impact performance. IIRC our implementation was using way too many allocations and instead of fixing that we decided, that from a practical standpoint it wasn’t worth the time when application is running 99,999% of the time with information level logging. Therefore, now only few key properties are logged and others were useful only while debugging issues during early development cycles). But it still would be nice to have baked in optimized option to include certain properties on log level switch.

**Comment by nblumhardt** [comment]
@vlm `.Enrich.AtLevel(LogEventLevel.Debug, levelSwitch, e => e.WithMachineName())` would mean the enricher would only apply when the `levelSwitch` is set to `Debug` or lower (e.g. `Verbose`).

Unless the level switch is used to connect the logger configuration to the extension method, we'd need to come up with some other kind of internal mechanism for the extension to learn the "final" minimum level - which might be more investment than we'd currently want to make in this feature (although not off the table for future consideration).

**Comment by nblumhardt** [comment]
#1316 introduces `Enrich.When()` as a catch-all way to implement this. I think the feature as originally envisioned in this PR is still worth considering, but the need to work with level switches and so-on makes the API potentially confusing.

Here's how it could be done:

```csharp
var levelSwitch = new LoggingLevelSwitch();

Log.Logger = new LoggerConfiguration()
    .MinimumLevel.ControlledBy(levelSwitch)
    .Enrich.When(_ => levelSwitch.MinimumLevel <= LogEventLevel.Information, e => e.WithMachineName())
    .WriteTo.Console()
    .CreateLogger();
```

Because `LoggerEnrichmentConfiguration.Wrap()` is public, the `.Enrich.WhenMinimumLevelIs(LogEventLevel.Debug, levelSwitch, e => e.WithMachineName())` type of extension could be written in a separate assembly, and could benefit from `<appSettings>`/`appSettings.json` support from level switches and actions.

👀 on the PR appreciated :-)

**Comment by andre-ss6** [comment]
Looks awesome to me! And btw, how would be appsettings configuration for this look like?

**Comment by nblumhardt** [comment]
@andre-ss6 the settings support for this would need to be similar to what we have today enabling `WriteTo.Async()`, `WriteTo.Logger()` and a few more scenarios. I haven't spent any time on the details, though.

**Comment by nblumhardt** [comment]
The above is now available in the 2.9.0-dev-* packages.
